File: backend/src/main/java/jupiterpi/vocabulum/webappserver/auth/AuthController.java
package jupiterpi.vocabulum.webappserver.auth;

import jupiterpi.vocabulum.core.db.Database;
import jupiterpi.vocabulum.core.users.User;
import jupiterpi.vocabulum.webappserver.auth.dtos.CredentialsDTO;
import jupiterpi.vocabulum.webappserver.auth.dtos.CredentialsVerificationDTO;
import jupiterpi.vocabulum.webappserver.auth.dtos.UserDetailsDTO;
import jupiterpi.vocabulum.webappserver.auth.registration.PendingRegistrations;
import jupiterpi.vocabulum.webappserver.auth.registration.RegistrationDTO;
import jupiterpi.vocabulum.webappserver.controller.CoreService;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.security.Principal;
import java.util.Optional;

@RestController
@RequestMapping("/auth")
public class AuthController {
    public AuthController() {
        CoreService.get();
    }

    public PendingRegistrations pendingRegistrations = new PendingRegistrations();

    @PostMapping("/register")
    public void register(@RequestBody RegistrationDTO dto) {
        pendingRegistrations.addPendingRegistration(dto);
    }

    @PostMapping("/confirmRegistration/{id}")
    public String confirmRegistration(@PathVariable String id) {
        User user = pendingRegistrations.confirmRegistration(id);
        if (user == null) throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Pending registration not found");
        return user.getName();
    }

    @PostMapping("/verifyCredentials")
    public CredentialsVerificationDTO verifyCredentials(@RequestBody CredentialsDTO dto) {
        Optional<User> userFound = Database.get().getUsers().getAll().stream()
                .filter(user -> user.getName().equals(dto.getUsername()) || user.getEmail().equals(dto.getUsername()))
                .filter(user -> user.getPassword().equals(dto.getPassword()))
                .findFirst();
        if (userFound.isEmpty()) {
            return new CredentialsVerificationDTO(false, "");
        } else {
            return new CredentialsVerificationDTO(true, userFound.get().getEmail());
        }
    }

    @PostMapping("/login")
    public UserDetailsDTO login(Principal principal) {
        User user = DbAuthenticationProvider.getUser(principal);
        return UserDetailsDTO.fromUser(user);
    }
}

File: backend/src/main/java/jupiterpi/vocabulum/webappserver/auth/AuthController.kt
package jupiterpi.vocabulum.webappserver.auth

import jupiterpi.vocabulum.webappserver.db.models.User
import jupiterpi.vocabulum.webappserver.db.models.Vouchers
import org.springframework.web.bind.annotation.*
import retrofit2.http.Path
import java.security.Principal
import java.util.*

@RestController
@RequestMapping("/auth")
class AuthController {
    @PostMapping("/register")
    fun register(@RequestHeader("Authorization") authHeader: String, @RequestBody dto: RegistrationDTO): UserDetailsDTO {
        val token = Auth.verifyAuthHeader(authHeader) ?: throw Exception("Invalid auth token")
        return UserDetailsDTO(User(token.uid, dto.username).also { it.save() })
    }
    data class RegistrationDTO(
        val username: String,
    )

    @GetMapping("/isAdmin")
    fun isAdmin(@RequestHeader("Authorization") authHeader: String) = IsAdminDTO(Auth.getToken(authHeader).claims["admin"] == true)
    data class IsAdminDTO(
        val isAdmin: Boolean,
    )

    @PostMapping("/makeAdmin/{uid}")
    fun makeAdmin(@RequestHeader("Authorization") authHeader: String, @PathVariable("uid") targetUid: String) {
        Auth.getToken(authHeader).assertAdmin()
        Auth.makeAdmin(targetUid, true)
    }
    @PostMapping("/revokeAdmin/{uid}")
    fun revokeAdmin(@RequestHeader("Authorization") authHeader: String, @PathVariable("uid") targetUid: String) {
        Auth.getToken(authHeader).assertAdmin()
        Auth.makeAdmin(targetUid, false)
    }

    @PostMapping("/useVoucher/{code}")
    fun useVoucher(@RequestHeader("Authorization") authHeader: String, @PathVariable code: String): UserDetailsDTO {
        val user = Auth.getUser(authHeader)
        val voucher = Vouchers.findByCode(code)
        if (voucher != null && !voucher.isExpired && !voucher.isUsed) {
            if (!user.isProUser) {
                voucher.useNow(user.completeKey)
                voucher.save()

                user.proExpiration = voucher.expiration
                user.save()
            }
        }
        return UserDetailsDTO(user)
    }
}

data class UserDetailsDTO(
    val username: String,
    val isProUser: Boolean,
    val discordUsername: String?,
) {
    constructor(user: User) : this(
        user.name,
        user.isProUser,
        user.discordUsername,
    )
}